libbeat/common/backoff.go

```diff
@@ -17,40 +17,44 @@
 
 package common
 
-import "time"
+import (
+	"math/rand"
+	"time"
+)
 
 // A Backoff waits on errors with exponential backoff (limited by maximum
 // backoff). Resetting Backoff will reset the next sleep timer to the initial
 // backoff duration.
 type Backoff struct {
-	duration time.Duration
-	done     <-chan struct{}
+	done <-chan struct{}
 
 	init time.Duration
 	max  time.Duration
 
+	tries int64
+
 	last time.Time
 }
 
 func NewBackoff(done <-chan struct{}, init, max time.Duration) *Backoff {
 	return &Backoff{
-		duration: init,
-		done:     done,
-		init:     init,
-		max:      max,
+		done: done,
+		init: init,
+		max:  max,
 	}
 }
 
 func (b *Backoff) Reset() {
-	b.duration = b.init
+	b.tries = 0
 }
 
+// Wait waits for a period of time based on the numbers of tries and a random jitter for the sleep.
 func (b *Backoff) Wait() bool {
-	backoff := b.duration
-	b.duration *= 2
-	if b.duration > b.max {
-		b.duration = b.max
-	}
+	b.tries++
+
+	// Make sure we have always some minimal back off and jitter even if we hit the cap.
+	temp := min64(int64(b.max), int64(b.init)*2*b.tries) / 2
+	backoff := time.Duration(temp + rand.Int63n(temp))
```

Comment:
This kind of looks like the 'full jitter' approach, but here the waiting times (besides the random part) only grow linearly. You need exponent -> `init * (1 << b.tries)`.

The former `b.duration` value is exactly the temp value here (assuming pow, not mult), as it's series goes like this:
- duration_0 = init = 2^0 * init
- duration_1 = 2 * duration_0 = 2 * init = 2^1 * init
- duration_2 = 2 * duration_1 = 2 * 2 * init = 2^2 * init
- ...


The problem with backoff by waiting is: No one knows when it is a good time to try again. And all clients act independent of each other, while influencing each other by binding resources on the server.

As we use backoff in a few places in Beats I wonder if we always want to use this approach, have the developer have a choice, or even make it fully configurable.

e.g.

```
type Backoff interface {
  Wait()
  Reset()
}

func WaitOnErr(b Backoff, err error) bool {
  ...
}

func TryWaitOnError(b Backoff, err error) bool {
  ...
}

func ExpBackoff(init, max time.Duration) Backoff {
  ...
}

func FullJitterBackoff(init, max time.Duration) Backoff {
  ...
}

func DecorrJitterBackoff(init, max time.Duration) Backoff {
  ...
}

func ConstWait(dur time.Duration) Backoff {
  ...
}

func BinBackoff(slots []time.Duration) Backoff {
  ...
}
```

If we want to add support for more scenarios, we should make it a separate package.

Reply:
:+1: for making this configurable. `common.Backoff` is used from file polling to reattempting to connect to outputs. Both use cases have different constraints. In case of file polling, it is not necessary to read new logs from files in random periods as only one Filebeat instance is trying to access it. However, when trying to connect to an external service (e.g. Logstash), it is required to have different backoff times to avoid overloading the service.

Reply:
@urso ah crap, forgot the *expo* 🤕, OK, I think I will do the following, create a special package adding an interface Backoff and keep the old one pointing to the new code.

@kvch Ah! You are absolutely right I did forget that we use the same logic in the file.